src/App.tsx

```diff
@@ -1,24 +1,220 @@
+      }
+    });
+  }, [filterOption, todos]);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then((todosFromServer) => {
+        setError(null);
+        setTodos(todosFromServer);
+      })
+      .catch(() => {
+        setError('Unable to upload todos');
+      });
+  }, []);
+
+  useEffect(() => {
+    let errorTimer: number;
+
+    if (error) {
+      errorTimer = window.setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+
+    return () => {
+      clearTimeout(errorTimer);
+    };
+  }, [error]);
+
+  const addTodo = useCallback(async (title: string) => {
+    setIsLoadingAddTodo(true);
+
+    const newTodo = {
+      id: 0,
+      completed: false,
+      title,
+      userId: USER_ID,
+    };
+
+    setTempTodo({ ...newTodo });
+
+    try {
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, createdTodo]);
+    } catch (err) {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+      setIsLoadingAddTodo(false);
+    }
+  }, []);
+
+  const removeTodo = useCallback((todoId: number) => {
+    setLoadingItems(prevItems => [...prevItems, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(
+          todo => todo.id !== todoId,
+        ));
+      })
+      .catch(() => {
+        setError('Unable to delete a todo');
+      })
+      .finally(() => {
+        setLoadingItems(prevItems => prevItems.filter(
+          id => id !== todoId,
+        ));
+      });
+  }, []);
+
+  const removeCompletedTodos = useCallback(() => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    completedTodos.forEach(todo => removeTodo(todo.id));
+  }, [removeTodo, todos]);
+
+  const editTodo = useCallback(async (
+    todoId: number,
+    args: UpdateTodoArgs,
+  ) => {
+    setLoadingItems(prevItems => [...prevItems, todoId]);
+
+    const updatedTodo: UpdateTodoArgs = { ...args };
+
+    try {
+      const response = await updateTodo(todoId, updatedTodo);
+
+      if (!response) {
+        return;
+      }
+
+      const updatedTodoData = response;
+
+      setTodos(prevTodos => prevTodos.map(todo => (
+        todo.id === todoId
+          ? { ...todo, ...updatedTodoData }
+          : todo
+      )));
+    } catch (err) {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingItems(prevItems => prevItems.filter(id => id !== todoId));
+    }
+  }, []);
+
+  const toggleTodosCompleted = useCallback(() => {
+    if (!isAllCompleted) {
+      todos.forEach(todo => {
+        if (!todo.completed) {
+          editTodo(
+            todo.id,
+            { completed: !isAllCompleted },
+          );
+        }
+      });
+    }
+
+    if (isAllCompleted) {
+      todos.forEach(todo => {
+        editTodo(
+          todo.id,
+          { completed: !isAllCompleted },
+        );
+      });
+    }
+  }, [editTodo, isAllCompleted, todos]);
```

Comment:
```suggestion
  const toggleTodosCompleted = useCallback(() => {
    todos.forEach(todo => {
        if (!todo.completed) {
          editTodo(
            todo.id,
            { completed: !isAllCompleted },
          );
        }
      });
  }, [editTodo, isAllCompleted, todos]);
```

shouldn't it work just fine if you write it like this?

Reply:
It doesn't work, in your way I can change todos only when them uncompleted, but can't change when they all completed. But I've tryed do something similar